lsp/nls/src/linearization/mod.rs

```diff
@@ -247,35 +260,72 @@ impl Linearizer for AnalysisHost {
                 x.push(ident.to_owned())
             }
             Term::MetaValue(meta) => {
-                // Notice 1: No push to lin
-                // Notice 2: we discard the encoded value as anything we
-                //           would do with the value will be handled in the following
-                //           call to [Self::add_term]
+                // DF walk AST (of Contracts)
+                // adds usages of user defined contracts to the linearization
+                fn walk_terms(
+                    lin: &mut Linearization<Building>,
+                    mut host: AnalysisHost,
+                    RichTerm { term, pos }: &RichTerm,
+                ) {
+                    host.add_term(lin, term, *pos, TypeWrapper::Concrete(AbsType::Dyn()));
+                    match &**term {
+                        Term::Op1(_, rt) => walk_terms(lin, host, rt),
+                        Term::Op2(_, rt1, rt2) => {
+                            walk_terms(lin, host.scope(), rt1);
+                            walk_terms(lin, host.scope(), rt2);
+                        }
+                        Term::OpN(_, rts) => {
+                            rts.iter().for_each(|rt| walk_terms(lin, host.scope(), rt))
+                        }
+                        Term::StrChunks(chunks) => chunks
+                            .iter()
+                            .filter_map(|chunk| match chunk {
+                                nickel::term::StrChunk::Literal(_) => None,
+                                nickel::term::StrChunk::Expr(rt, _) => Some(rt),
+                            })
+                            .for_each(|rt| walk_terms(lin, host.scope(), rt)),
+                        Term::App(rt1, rt2) => {
+                            walk_terms(lin, host.scope(), rt1);
+                            walk_terms(lin, host.scope(), rt2);
+                        }
+                        Term::Record(rts, _) | Term::RecRecord(rts, _, _) => rts
```

Comment:
Do we need to walk dynamic fields?

Reply:
For stupid Contracts like 
```
... | #({some.contract = Contract}.some.contract)
```

or what other kind of dynamic fields do you mean?

Reply:
I'm not sure if it would come up in contracts.
I'm wondering if we need to traverse the second field of `RecRecord`.
```rust
    RecRecord(
        HashMap<Ident, RichTerm>,
        Vec<(RichTerm, RichTerm)>, /* field whose name is defined by interpolation */
        RecordAttrs,
    ),
```

Reply:
Oh, yeah they are not typechecked so i never dealt with those guys
As the TC doesnt touch them either, we should do so in the linearizer which would mean pulling up this walk method to be used by the linearizer.
I'm not sure if i want to do that here or in a later PR

Reply:
Either is fine